edx_proctoring/api.py

```diff
@@ -0,0 +1,558 @@
+    exam = get_exam_by_id(exam_id)
+    allowed_time_limit_mins = exam['time_limit_mins']
+
+    # add in the allowed additional time
+    allowance = ProctoredExamStudentAllowance.get_allowance_for_user(
+        exam_id,
+        user_id,
+        "Additional time (minutes)"
+    )
+
+    if allowance:
+        allowance_extra_mins = int(allowance.value)
+        allowed_time_limit_mins += allowance_extra_mins
+
+    attempt_code = unicode(uuid.uuid4()).upper()
+
+    external_id = None
+    if taking_as_proctored:
+        callback_url = 'http://{hostname}{path}'.format(
+            hostname=settings.SITE_NAME,
+            path=reverse(
+                'edx_proctoring.anonymous.proctoring_launch_callback.start_exam',
+                args=[attempt_code]
+            )
+        )
+
+        # now call into the backend provider to register exam attempt
+        external_id = get_backend_provider().register_exam_attempt(
+            exam,
+            allowed_time_limit_mins,
+            attempt_code,
+            False,
+            callback_url
+        )
+
+    attempt = ProctoredExamStudentAttempt.create_exam_attempt(
+        exam_id,
+        user_id,
+        '',  # student name is TBD
+        allowed_time_limit_mins,
+        attempt_code,
+        taking_as_proctored,
+        False,
+        external_id
+    )
+    return attempt.id
+
+
+def start_exam_attempt(exam_id, user_id):
+    """
+    Signals the beginning of an exam attempt for a given
+    exam_id. If one already exists, then an exception should be thrown.
+
+    Returns: exam_attempt_id (PK)
+    """
+
+    existing_attempt = ProctoredExamStudentAttempt.objects.get_exam_attempt(exam_id, user_id)
+
+    if not existing_attempt:
+        err_msg = (
+            'Cannot start exam attempt for exam_id = {exam_id} '
+            'and user_id = {user_id} because it does not exist!'
+        ).format(exam_id=exam_id, user_id=user_id)
+
+        raise StudentExamAttemptDoesNotExistsException(err_msg)
+
+    return _start_exam_attempt(existing_attempt)
+
+
+def start_exam_attempt_by_code(attempt_code):
+    """
+    Signals the beginning of an exam attempt when we only have
+    an attempt code
+    """
+
+    existing_attempt = ProctoredExamStudentAttempt.objects.get_exam_attempt_by_code(attempt_code)
+
+    if not existing_attempt:
+        err_msg = (
+            'Cannot start exam attempt for attempt_code = {attempt_code} '
+            'because it does not exist!'
+        ).format(attempt_code=attempt_code)
+
+        raise StudentExamAttemptDoesNotExistsException(err_msg)
+
+    return _start_exam_attempt(existing_attempt)
+
+
+def _start_exam_attempt(existing_attempt):
+    """
+    Helper method
+    """
+
+    if existing_attempt.started_at:
+        # cannot restart an attempt
+        err_msg = (
+            'Cannot start exam attempt for exam_id = {exam_id} '
+            'and user_id = {user_id} because it has already started!'
```

Comment:
@chrisndodge I'm not sure whether these error message are shown to students. If yes, I will provide some suggested rewording.